src/librustc_errors/emitter.rs

```diff
@@ -31,6 +31,11 @@ const ANONYMIZED_LINE_NUM: &str = "LL";
 pub trait Emitter {
     /// Emit a structured diagnostic.
     fn emit(&mut self, db: &DiagnosticBuilder);
+
+    /// Shows explanation about "rustc --explain"
+    fn show_explain(&mut self) -> Vec<String> {
```

Comment:
While this does successfully implement #48041, it seems rather _ad hoc_ to make a `show_explain` method be part of the contract of the `Emitter` trait. (In contrast to how it makes sense for all `Emitter`s to have an `emit` method, because that's what being an `Emitter` is all about.) Can we implement this in a more generalizable way?

Two potential alternatives that come to mind:
 * Make the `Emitter` trait have a `destination` method that exposes something we can write arbitrary lines to, and use it to write the `--explain` usage message in `abort_if_errors`.
   * `JsonEmitter`'s `dst` field is of type `Box<Write + Send>`, and `EmitterWriter`'s `Destination` implements `Write` and the field of all three of its variants implement `Send` (visible [in the source](https://github.com/rust-lang/rust/blob/9cb18a92ad87852c4c5d6726b8fbe8c38deda4ba/src/librustc_errors/emitter.rs#L1479-L1485) in the case of `BufferedTerminal` and `Raw`, and the [`term` docs say that](https://stebalien.github.io/doc/term/term/type.StderrTerminal.html) `StderrTerminal` is `Send`), so I _think_ the types should check out.
 * Implement the `--explain` message as a `Diagnostic`, adding a plainer formatting mode to `Diagnostic` if necessary.

Reply:
@GuillaumeGomez can you comment on this?